internal/provider/data_source_gitlab_branch_test.go

```diff
@@ -0,0 +1,110 @@
+package provider
+
+import (
+	"fmt"
+	"strconv"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/acctest"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
+)
+
+func TestAccDataGitlabBranch_basic(t *testing.T) {
+	rInt := acctest.RandInt()
+	branchProtection := testAccGitlabBranchProtectionAttributes{}
+	resource.Test(t, resource.TestCase{
+		PreCheck:          func() { testAccPreCheck(t) },
+		ProviderFactories: providerFactories,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccDataGitlabBranch(rInt),
+				Check: resource.ComposeTestCheckFunc(
+					testAccDataSourceGitlabBranch("gitlab_branch.foo", "data.gitlab_branch.foo"),
+					testAccDataSourceGitlabBranchProtection("data.gitlab_branch.foo", branchProtection),
+				),
+			},
+		},
+	})
+}
+
+func testAccDataSourceGitlabBranch(src, n string) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+
+		branch := s.RootModule().Resources[src]
+		branchAttr := branch.Primary.Attributes
+
+		search := s.RootModule().Resources[n]
+		searchAttr := search.Primary.Attributes
+
+		testAttributes := []string{
+			"id",
+			"name",
+			"web_url",
+			"default",
+			"project",
+			"can_push",
+			"merged",
+			"commit",
+			"parent_ids",
+		}
+
+		for _, attribute := range testAttributes {
+			if searchAttr[attribute] != branchAttr[attribute] {
+				return fmt.Errorf("expected branch's parameter `%s` to be: %s, but got: `%s`", attribute, branchAttr[attribute], searchAttr[attribute])
+			}
+		}
+		return nil
+	}
+}
+
+func testAccDataSourceGitlabBranchProtection(n string, exp testAccGitlabBranchProtectionAttributes) resource.TestCheckFunc {
```

Comment:
What's the idea of having a separate function for the branch protection fields?

Reply:
Without it I get this error:
https://github.com/gitlabhq/terraform-provider-gitlab/runs/5274065200?check_suite_focus=true#step:6:8

Reply:
It was a bug because I initially didn't have the protection fields on the resource so the lookup between datasource : resource in this test was not matching for those fields -- added this to the resource now and regenerated docs